Make sure tablets with WALs are not merged

cshannon opened the pull request:
Add extra verification to make sure that no tablets that are merged contain any WALs to prevent losing data

This fixes #3845 for the 2.1 branch. For main there will be some slight modifications due to no-chopped merge but the fix is mostly going to be the same and I can fix conflicts on the merge.

cshannon commented:
@keith-turner - I moved some of the new logic into a couple package level methods to add a couple tests but most of this stuff is very hard to test without spending a ton of time mocking or setting up ITs. I figure the existing merge and delete ITs should be able to check things.

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/state/MergeStats.java:
```diff
@@ -271,6 +281,22 @@ private boolean verifyMergeConsistency(AccumuloClient accumuloClient, CurrentSta
 
+  @VisibleForTesting
+  boolean verifyState(MergeInfo info) {
+    if (info.getState() != MergeState.WAITING_FOR_OFFLINE) {
+      return false;
+    }
+    return true;
```
Another merge state seems really unexpected here, wonder if an exception would be better.

```suggestion
    Preconditions.checkstate(info.getState() == MergeState.WAITING_FOR_OFFLINE, "Unexpected merge state %", info.getState());
    if (info.getState() != MergeState.WAITING_FOR_OFFLINE) {
      return false;
    }
    return true;
```

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/state/MergeStats.java:
```diff
@@ -271,6 +281,22 @@ private boolean verifyMergeConsistency(AccumuloClient accumuloClient, CurrentSta
+
+  @VisibleForTesting
+  boolean verifyWalogs(TabletLocationState tls) {
+    if (!tls.walogs.isEmpty()) {
+      return false;
+    }
+    return true;
```
```suggestion
    return tls.walogs.isEmpty();
```

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -783,13 +784,24 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
-        if (key.getColumnFamily().equals(DataFileColumnFamily.NAME)) {
+
+        // Verify that Tablet is offline
+        if (key.getColumnFamily().equals(CurrentLocationColumnFamily.NAME)) {
```
Can also check for a future location, should not see that either.

```suggestion
        if (key.getColumnFamily().equals(CurrentLocationColumnFamily.NAME)  || key.getColumnFamily().equals(FutureLocationColumnFamily.NAME)) {
```

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:
```diff
@@ -783,13 +784,24 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
       ServerColumnFamily.DIRECTORY_COLUMN.fetch(scanner);
+
       scanner.fetchColumnFamily(DataFileColumnFamily.NAME);
+      scanner.fetchColumnFamily(CurrentLocationColumnFamily.NAME);
```
```suggestion
      scanner.fetchColumnFamily(CurrentLocationColumnFamily.NAME);
      scanner.fetchColumnFamily(FutureLocationColumnFamily.NAME);
```

keith-turner reviewed (commented):
Could also add validation to the deleteTablets code in TGW.  Seems its already checking for current location and throwing an exception.  Could add checks for wals and future locations.

cshannon commented:
@keith-turner - I added the requested changes but one thing I was just thinking is that I added the WALs check to the deleteTablets code but we shouldn't really care about that if we are deleting the tablets anyways.  We would only care about the tablets at the start/end of the delete that are being kept so not sure if I need to keep the check. It probably doesn't hurt to keep it so it's in a consistent state and we don't have to deal with edge cases around start/end range. Things are also a bit different depending on the version. In 2.1 obviously tablets chopped and splitt at the deletion range and whole tablets are gone but things work different with no chop merge and fencing (fence start/end tablet without splitting) and in elasticity (fence and merge empty tablets)

keith-turner commented on test/src/main/java/org/apache/accumulo/test/manager/MergeStateIT.java:
```diff
@@ -207,6 +208,24 @@ public void test() throws Exception {
           Location.current(state.someTServer), null, null, walogs, false)), null);
 
+      // Add a walog which should keep the state from transitioning to MERGING
+      KeyExtent ke = new KeyExtent(tableId, new Text("t"), new Text("p"));
```
This is a nice test for something that is really tricky to test.
